zipkin-storage/zipkin2_cassandra/src/main/java/zipkin2/storage/cassandra/internal/call/AccumulateAllResults.java

```diff
@@ -0,0 +1,70 @@
+/**
+ * Copyright 2015-2017 The OpenZipkin Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package zipkin2.storage.cassandra.internal.call;
+
+import com.datastax.driver.core.ResultSet;
+import com.datastax.driver.core.Row;
+import com.google.auto.value.AutoValue;
+import com.google.common.util.concurrent.ListenableFuture;
+import java.util.function.BiConsumer;
+import java.util.function.Supplier;
+import zipkin2.Call;
+import zipkin2.Call.FlatMapper;
+
+public abstract class AccumulateAllResults<T> implements FlatMapper<ResultSet, T> {
+  protected abstract Supplier<T> supplier();
+
+  protected abstract BiConsumer<Row, T> accumulator();
+
+  @Override public Call<T> map(ResultSet rs) {
+    return new AutoValue_AccumulateAllResults_AccumulateNextResults<>(
+      supplier().get(),
+      accumulator()
+    ).map(rs);
+  }
+
+  @AutoValue
+  static abstract class FetchMoreResults extends ListenableFutureCall<ResultSet> {
+    abstract ResultSet resultSet();
+
+    @Override protected ListenableFuture<ResultSet> newFuture() {
+      return resultSet().fetchMoreResults();
+    }
+
+    @Override public Call<ResultSet> clone() {
+      throw new UnsupportedOperationException();
+    }
+  }
+
+  @AutoValue
+  static abstract class AccumulateNextResults<T> implements FlatMapper<ResultSet, T> {
+    abstract T pendingResults();
+
+    abstract BiConsumer<Row, T> accumulator();
+
+    @Override public Call<T> map(ResultSet rs) {
+      if (!rs.isFullyFetched()) rs.fetchMoreResults(); // TODO: dropped future
```

Comment:
@llinder if you know why we can drop futures in the places I made TODO, or why 2000 is special, please note and I'll backfill some comments.

Reply:
`rs.fetchMoreResult()` is both a hint as well as a way to wait on the subsequent page.
here it is only being used as a hint: that is to initiate in the background the fetch of the subsequent page while we process the current page. otherwise the fetch for the next page is not initiated until this page becomes exhausted. during the processing of any specific page, multiple calls to `rs.fetchMoreResult()` will always return the same future. even if we do nothing with the returned future, when the current page is exhausted it will seamlessly move onto the next page as it has been fetched for us in the background (asynchronously).

`2000` is just an appropriate place (through the default of 5000 rows per page) to pre-emptively kick-off the asynchronous fetch of the next page.

Reply:
thanks. will refactor slightly to ensure call modeling is coherent with this